File: src/main/java/me/miki/shindo/management/settings/impl/ColorSetting.java
package me.miki.shindo.management.settings.impl;

import me.miki.shindo.management.settings.config.ConfigOwner;
import me.miki.shindo.management.language.TranslateText;
import me.miki.shindo.management.settings.Setting;
import me.miki.shindo.utils.ColorUtils;

import java.awt.*;

public class ColorSetting extends Setting {

    private final Color defaultColor;
    private final boolean showAlpha;
    private float hue, saturation, brightness;
    private int alpha;
    private Color color;

    public ColorSetting(TranslateText text, ConfigOwner parent, Color color, boolean showAlpha) {
        super(text, parent);

        this.color = color;
        this.defaultColor = color;
        this.hue = ColorUtils.getHue(color);
        this.saturation = ColorUtils.getSaturation(color);
        this.brightness = ColorUtils.getBrightness(color);
        this.alpha = color.getAlpha();
        this.showAlpha = showAlpha;
    }

    public ColorSetting(String name, ConfigOwner parent, Color color, boolean showAlpha) {
        super(name, parent);

        this.color = color;
        this.defaultColor = color;
        this.hue = ColorUtils.getHue(color);
        this.saturation = ColorUtils.getSaturation(color);
        this.brightness = ColorUtils.getBrightness(color);
        this.alpha = color.getAlpha();
        this.showAlpha = showAlpha;
    }

    @Override
    public void reset() {
        this.color = defaultColor;
        this.hue = ColorUtils.getHue(color);
        this.saturation = ColorUtils.getSaturation(color);
        this.brightness = ColorUtils.getBrightness(color);
        this.alpha = color.getAlpha();
    }

    public Color getColor() {
        return color;
    }

    public void setColor(Color color) {
        this.color = color;
    }

    public Color getDefaultColor() {
        return defaultColor;
    }

    public float getHue() {
        return hue;
    }

    public void setHue(float hue) {
        this.hue = hue;
        this.color = ColorUtils.applyAlpha(Color.getHSBColor(hue, saturation, brightness), alpha);
    }

    public float getSaturation() {
        return saturation;
    }

    public void setSaturation(float saturation) {
        this.saturation = saturation;
        this.color = ColorUtils.applyAlpha(Color.getHSBColor(hue, saturation, brightness), alpha);
    }

    public float getBrightness() {
        return brightness;
    }

    public void setBrightness(float brightness) {
        this.brightness = brightness;
        this.color = ColorUtils.applyAlpha(Color.getHSBColor(hue, saturation, brightness), alpha);
    }

    public int getAlpha() {
        return alpha;
    }

    public void setAlpha(int alpha) {
        this.alpha = alpha;
        this.color = ColorUtils.applyAlpha(color, alpha);
    }

    public boolean isShowAlpha() {
        return showAlpha;
    }
}

File: src/main/kotlin/com/shindoclient/shindo/management/settings/impl/ColorSetting.kt
package com.shindoclient.shindo.management.settings.impl

import com.shindoclient.shindo.management.language.TranslateText
import com.shindoclient.shindo.management.settings.Setting
import com.shindoclient.shindo.management.settings.config.ConfigOwner
import com.shindoclient.shindo.utils.ColorUtils
import java.awt.Color

open class ColorSetting : Setting {
    private val defaultColor: Color
    private val showAlpha: Boolean
    private var hue: Float
    private var saturation: Float
    private var brightness: Float
    private var alpha: Int
    private var color: Color

    constructor(text: TranslateText, parent: ConfigOwner, color: Color, showAlpha: Boolean) : super(text, parent) {
        this.color = color
        this.defaultColor = color
        this.hue = ColorUtils.getHue(color)
        this.saturation = ColorUtils.getSaturation(color)
        this.brightness = ColorUtils.getBrightness(color)
        this.alpha = color.alpha
        this.showAlpha = showAlpha
    }

    constructor(name: String, parent: ConfigOwner, color: Color, showAlpha: Boolean) : super(name, parent) {
        this.color = color
        this.defaultColor = color
        this.hue = ColorUtils.getHue(color)
        this.saturation = ColorUtils.getSaturation(color)
        this.brightness = ColorUtils.getBrightness(color)
        this.alpha = color.alpha
        this.showAlpha = showAlpha
    }

    override fun reset() {
        color = defaultColor
        hue = ColorUtils.getHue(color)
        saturation = ColorUtils.getSaturation(color)
        brightness = ColorUtils.getBrightness(color)
        alpha = color.alpha
    }

    fun getColor(): Color = color

    open fun setColor(color: Color) {
        this.color = color
    }

    fun getDefaultColor(): Color = defaultColor

    fun getHue(): Float = hue

    open fun setHue(hue: Float) {
        this.hue = hue
        color = ColorUtils.applyAlpha(Color.getHSBColor(hue, saturation, brightness), alpha)
    }

    fun getSaturation(): Float = saturation

    open fun setSaturation(saturation: Float) {
        this.saturation = saturation
        color = ColorUtils.applyAlpha(Color.getHSBColor(hue, saturation, brightness), alpha)
    }

    fun getBrightness(): Float = brightness

    open fun setBrightness(brightness: Float) {
        this.brightness = brightness
        color = ColorUtils.applyAlpha(Color.getHSBColor(hue, saturation, brightness), alpha)
    }

    fun getAlpha(): Int = alpha

    open fun setAlpha(alpha: Int) {
        this.alpha = alpha
        color = ColorUtils.applyAlpha(color, alpha)
    }

    fun isShowAlpha(): Boolean = showAlpha
}
